src/hl/group.rs

```diff
@@ -1,17 +1,24 @@
 use std::fmt::{self, Debug};
 use std::ops::Deref;
+use std::panic;
 
 use hdf5_sys::{
     h5::{hsize_t, H5_index_t, H5_iter_order_t},
     h5d::H5Dopen2,
     h5g::{H5G_info_t, H5Gcreate2, H5Gget_info, H5Gopen2},
-    h5l::{
-        H5L_info1_t as H5L_info_t, H5L_iterate1_t as H5L_iterate_t, H5Lcreate_hard, H5Lcreate_soft,
-        H5Ldelete, H5Lexists, H5Literate1 as H5Literate, H5Lmove, H5L_SAME_LOC,
-    },
+    h5l::{H5Lcreate_hard, H5Lcreate_soft, H5Ldelete, H5Lexists, H5Lmove, H5L_SAME_LOC},
     h5p::{H5Pcreate, H5Pset_create_intermediate_group},
 };
 
+#[cfg(not(hdf5_1_12_0))]
+use hdf5_sys::h5l::{
```

Comment:
Hmm, so that's what I was mentioning earlier, maybe we're doing it all backwards? 

```rust
    #[cfg_attr(hdf5_1_12_0, deprecated(note = "deprecated in HDF5 1.12.0, use H5Literate2()"))]
    #[cfg_attr(not(hdf5_1_12_0), link_name = "H5Literate")]
    pub fn H5Literate1(
        grp_id: hid_t, idx_type: H5_index_t, order: H5_iter_order_t, idx: *mut hsize_t,
        op: H5L_iterate1_t, op_data: *mut c_void,
    ) -> herr_t;
    #[cfg(hdf5_1_12_0)]
    pub fn H5Literate2(
        grp_id: hid_t, idx_type: H5_index_t, order: H5_iter_order_t, idx: *mut hsize_t,
        op: H5L_iterate2_t, op_data: *mut c_void,
    ) -> herr_t;

    // Mimic HDF5 C header macros?
    #[cfg(not(hdf5_1_12_0))]
    pub use self::H5Literate1 as H5Literate;
    #[cfg(hdf5_1_12_0)]
    pub use self::H5Literate2 as H5Literate;
```

Then in `hdf5::hl::group` you just `use hdf5_sys::h5l::H5Literate`?

Reply:
The only confusing bit would be that C users might expect a different version, as the macro conditionally uses a particular version, which might not be the latest one

Reply:
Hm?

So, pre-1.12:

- If you use `H5Literate()`, it would resolve to the same in Rust/C
- We also have `H5Literate1` which resolves to the `H5Literate()`
- `H5Literate2()` doesn't exist in either

Post 1.12:

- `H5Literate2()` resolves to the same thing in Rust/C
- `H5Literate1()` resolves to the same thing in Rust/C
- We also have `H5Literate` which resolves to `H5Literate2()`

Reply:
A crate could directly use `H5Literate1` and remain forwards compatible, although with deprecation warnings. It would probably be a good idea to skim through the crate to make sure we have such aliasing for all structs and functions. This can be left for a future PR

Reply:
The `H5version.h` crate is responsible for selecting the API version of each function based on compile time flags, but we bypass this entirely and always use the latest API.

Reply:
Yea, right. I don't think there's an easy way out and in fact what we're doing is probably a clearer way to go as long as we do it with all 1/2/3 versioned structs and functions, tag them all with deprecation warnings and provide aliases to select one automatically like the one you added. As you've said, probably a separate PR to revise all existing versioned bindings, mind opening a ticket?

Reply:
#114